### PR TITLE
Rewrite indexing

### PR DESCRIPTION
**Pull request by xieydd**
@gaocegege @VoVAllen Any suggestions?

### CODE REVIEW

File: src/usage/indexing.md [hunk 1]
```diff
@@ -150,3 +59,29 @@ WITH (options = $$
 segment.max_growing_segment_size = 200000
 $$);
 ```
+
+## Suggestions
+
+Choosing the right indexing algorithm and options is important for the performance of the vector search. Here are some suggestions for different scenarios.
+
+### Brute force(FLAT)
```
Comment (kemingy) [inline src/usage/indexing.md:67]:
```suggestion
### Brute force (flat)
```
1. requires a space here (please also change others)
2. flat is not an abbr.

File: src/usage/indexing.md [hunk 2]
```diff
@@ -150,3 +59,29 @@ WITH (options = $$
...
+
+If you only perform a few searches(say 1000-10000) or you need guaranteed exact results, you can use the [brute force](/reference/indexing_options.html#options-for-table-flat) algorithm.
```
Comment (kemingy) [inline src/usage/indexing.md:69]: "a few searches" is ambiguous. It depends on the candidate size, not the query size.

File: src/usage/indexing.md [hunk 3]
```diff
@@ -150,3 +59,29 @@ WITH (options = $$
...
+
+It provides the baseline for results for the other indexes. It does not compress the vectors, but does not add overhead on top of them. It's simple and easy to use. However, it's not suitable for large datasets.
```
Comment (kemingy) [inline src/usage/indexing.md:71]:
```suggestion
It provides the baseline for the other indexes. It does not compress the vectors, but does not add overhead on top of them. It's simple and easy to use. However, it's not suitable for large datasets.
```

Please try the GPT polishing.

**Comment by xieydd** [comment]
@kemingy Thank you for your review. Do you have any suggestions regarding the architecture?

**Comment by kemingy** [comment]
> @kemingy Thank you for your review. Do you have any suggestions regarding the architecture?

Do you mean the file structure here? LGTM. Remember to update the related links to the original page.

**Review by usamoi — APPROVED** [review]

File: src/usage/indexing.md (hunk 2 above)
Comment (xieydd) [reply src/usage/indexing.md:69]: This means that if you run only a few queries, the time spent building indexes won't be offset by faster search times.

File: src/usage/indexing.md [hunk 4]
```diff
@@ -150,3 +59,29 @@ WITH (options = $$
...
+
+### Brute force (flat)
+
+If you only perform a few searches (say 1000-10000) or you need guaranteed exact results, you can use the [brute force](/reference/indexing_options.html#options-for-table-flat) algorithm.
+
+It provides the baseline for the other indexes. It does not compress the vectors, but does not add overhead on top of them. It's simple and easy to use. However, it's not suitable for large datasets.
+
+You can use the quantization to reduce the memory usage. See the [reference](/reference/indexing_options.html#options-for-table-quantization) for the full list of options. If you want to find details about the quantization in the [reference](/usage/quantization.html).
+
+### Inverted file index(IVF)
```
Comment (kemingy) [inline src/usage/indexing.md:75]:
```suggestion
### Inverted file index (IVF)
```

File: src/usage/indexing.md [hunk 5]
```diff
@@ -150,3 +59,29 @@ WITH (options = $$
...
+
+IVF mainly uses the idea of inverted indexing to store the vectors `(id, vector)` under each cluster center. When querying a vector, it finds the nearest several centers and searches for the vectors under these centers respectively. 
+
+Before searching, you need to train `nlist` cluster centers. And you can set `nsample` for K-Means clustering. The larger the value, the more accurate the clustering, but the longer the training time. Other options can be found in the [reference](/reference/indexing_options.html#options-for-table-ivf).
+
+### Hierarchical navigable small world graph(HNSW)
```
Comment (kemingy) [inline src/usage/indexing.md:81]:
```suggestion
### Hierarchical navigable small world graph (HNSW)
```

File: src/usage/indexing.md [hunk 6]
```diff
@@ -150,3 +59,29 @@ WITH (options = $$
...
+
+Choosing the right indexing algorithm and options is important for the performance of the vector search. Here are some suggestions for different scenarios.
+
+### Brute force (flat)
+
+If you only perform a few searches (say 1000-10000) or you need guaranteed exact results, you can use the [brute force](/reference/indexing_options.html#options-for-table-flat) algorithm.
+
+It provides the baseline for the other indexes. It does not compress the vectors, but does not add overhead on top of them. It's simple and easy to use. However, it's not suitable for large datasets.
+
+You can use the quantization to reduce the memory usage. See the [reference](/reference/indexing_options.html#options-for-table-quantization) for the full list of options. If you want to find details about the quantization in the [reference](/usage/quantization.html).
```
Comment (kemingy) [inline src/usage/indexing.md:73]: The last sentence is weird.

File: src/usage/indexing.md [hunk 7]
```diff
@@ -150,3 +59,29 @@ WITH (options = $$
...
+
+### Inverted file index(IVF)
+
+IVF mainly uses the idea of inverted indexing to store the vectors `(id, vector)` under each cluster center. When querying a vector, it finds the nearest several centers and searches for the vectors under these centers respectively. 
+
+Before searching, you need to train `nlist` cluster centers. And you can set `nsample` for K-Means clustering. The larger the value, the more accurate the clustering, but the longer the training time. Other options can be found in the [reference](/reference/indexing_options.html#options-for-table-ivf).
+
+### Hierarchical navigable small world graph(HNSW)
+
+HNSW combines the idea of [skip list](https://brilliant.org/wiki/skip-lists/) and navigable small world(NSW) graphs. Through layered format, longer edges in the highest layers (for fast search) and shorter edges in the lower layers (for accurate search).
```
Comment (kemingy) [inline src/usage/indexing.md:83]:
```suggestion
HNSW combines the idea of [skip list](https://brilliant.org/wiki/skip-lists/) and navigable small world (NSW) graphs. Through layered format, longer edges in the highest layers (for fast search) and shorter edges in the lower layers (for accurate search). 
```

**Review by kemingy — CHANGES_REQUESTED** [review]
The sentences in "indexing" need to be polished.

File: src/usage/indexing.md [hunk 8]
```diff
@@ -150,3 +79,51 @@ WITH (options = $$
...
+
+Before you begin your search, it's necessary to train nlist cluster centers. You also have the option to set nsample for K-Means clustering; a higher value yields more precise clusters but increases training time. For additional settings, please consult the reference guide.
+
+Before searching, train `nlist` cluster centers using K-Means clustering with `nsample` points. A higher value yields more precise clustering but requires longer training time. For additional settings, please consult the [reference guide](/reference/indexing_options.html#options-for-table-ivf).
+
+### Hierarchical navigable small world graph (HNSW)
+
+HNSW merges the concepts of [skip list](https://brilliant.org/wiki/skip-lists/) with navigable small world (NSW) graphs, employing a hierarchical structure that features longer edges on higher layers for rapid searching and shorter edges on lower layers to enhance search precision. 
+
+#### Navigable Small World
```
Comment (gaocegege) [inline src/usage/indexing.md:106]: We only capitalize the first letter in the doc page. Please:

```suggestion
#### Navigable small world
```

Comment (xieydd) [reply src/usage/indexing.md:106]: lgtm

**Comment by gaocegege** [comment]
LGTM

cc @kemingy

File: src/usage/indexing.md [hunk 9]
```diff
@@ -150,3 +79,51 @@ WITH (options = $$
...
+
+IVF mainly uses the idea of inverted indexing to store the vectors `(id, vector)` under each cluster center. When querying a vector, it finds the nearest several centers and searches for the vectors under these centers respectively.
+
+Before you begin your search, it's necessary to train nlist cluster centers. You also have the option to set nsample for K-Means clustering; a higher value yields more precise clusters but increases training time. For additional settings, please consult the reference guide.
+
+Before searching, train `nlist` cluster centers using K-Means clustering with `nsample` points. A higher value yields more precise clustering but requires longer training time. For additional settings, please consult the [reference guide](/reference/indexing_options.html#options-for-table-ivf).
+
+### Hierarchical navigable small world graph (HNSW)
+
+HNSW merges the concepts of [skip list](https://brilliant.org/wiki/skip-lists/) with navigable small world (NSW) graphs, employing a hierarchical structure that features longer edges on higher layers for rapid searching and shorter edges on lower layers to enhance search precision.
```
Comment (kemingy) [inline src/usage/indexing.md:104]: Use Wikipedia links https://en.wikipedia.org/wiki/Skip_list since it has a higher score.

Comment (xieydd) [reply src/usage/indexing.md:104]: sgtm

**Pull request merged**